[cdac] cdac-build-tool

lambdageek opened the pull request:
Contributes to https://github.com/dotnet/runtime/issues/99298

# cDAC Build Tool

## Summary

The purpose of `cdac-build-tool` is to generate a `.c` file that contains a JSON cDAC contract descriptor.

It works by processing one or more object files containing data descriptors and zero or more text
files that specify contracts.

## Running

```console
% cdac-build-tool compose [-v] -o contractdescriptor.c -c contracts.txt datadescriptor.o
```
## .NET runtime build integration

`cdac-build-tool` is meant to run as a CMake custom command.
It consumes a target platform object file and emits a C source
file that contains a JSON contract descriptor.  The C source
is the included in the normal build and link steps to create the runtime.

The contract descriptor source file depends on `contract-aux-data.c` which is a source file that contains
the definitions of the "indirect pointer data" that is referenced by the data descriptor.  This is typically the addresses of important global variables in the runtime.
Constants and build flags are embedded directly in the JSON payload.

Multiple data descriptor source files may be specified (for example if they are produced by different components of the runtime, or by different source languages).  The final JSON payload will be a composition of all the data descriptors.

Multiple contracts text files may be specified.  This may be useful if some contracts are conditionally included (for example if they are platform-specific).  The final JSON payload will be a composition of all the contracts files.

```mermaid
flowchart TB
  headers("runtime headers")
  data_header("datadescriptor.h")
  data_src("datadescriptor.c")
  compile_data["clang"]
  data_obj("datadescriptor.o")
  contracts("contracts.txt")
  globals("contractpointerdata.c")
  build[["cdac-build-tool"]]
  descriptor_src("contractdescriptor.c")
  vm("runtime sources")
  compile_runtime["clang"]
  runtime_lib(["libcoreclr.so"])

  headers -.-> data_src
  headers ~~~ data_header
  data_header -.-> data_src
  headers -.-> globals
  headers -.-> vm
  data_src --> compile_data --> data_obj --> build
  contracts ---> build
  build --> descriptor_src
  descriptor_src --> compile_runtime
  data_header -.-> globals ----> compile_runtime
  vm ----> compile_runtime --> runtime_lib
```

lambdageek committed:
checkpoint

lambdageek committed:
add old tooling notes

lambdageek committed:
add an emitter

lambdageek committed:
checkpoint read in the directory header

lambdageek committed:
checkpoint plausible-looking offsets

lambdageek committed:
checkpoint: field stuff is working

lambdageek committed:
checkpoint json

lambdageek committed:
contract parsing

lambdageek committed:
refactor

lambdageek committed:
indirect pointer value support

lambdageek committed:
move sample to tool dir

lambdageek committed:
Take baselines from the docs/design/datacontracts/data dir

We don't parse them yet, however

lambdageek committed:
Add README

lambdageek committed:
fix BE

Store the magic as a uint64_t so that it will follow the platform
endianness.

Store endmagic as bytes so that it directly follows the name pool -
and fix the endmagic check not to look at the endianness

lambdageek committed:
hook up cdac-build-tool to the coreclr build; export DotNetRuntimeContractDescriptor

lambdageek committed:
try fix win-x86 build warning

lambdageek committed:
paths

lambdageek committed:
vebose

lambdageek committed:
fix typo; fixes win32

lambdageek committed:
cleanup; add contracts.txt

lambdageek committed:
add diagram to README

lambdageek committed:
move implementation notes

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/data-descriptor.md:
```diff
@@ -0,0 +1,177 @@
+
+### Summary
+
+This is an internal implemetnation detail allowing tooling to read target architecture structure sizes and offsets without understaing target architecture object formats.
```
```suggestion
This is an internal implemetnation detail allowing tooling to read target architecture structure sizes and offsets without understanding target architecture object formats.
```

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ObjectFileScraper.cs:
```diff
@@ -0,0 +1,425 @@
+            int bytesRead = await stream.ReadAsync(buf, 0, buf.Length, token);
+            if (bytesRead == 0)
+                return new (){Found = false, Position = 0, LittleEndian = true};
+            // FIXME: what if magic spans a buffer boundary
```
This sort of tool is meant to process a single object file artifact. As such I think it makes sense to just read in the entire binary blob and entirely avoid these sorts of issues. I can't imagine any needed object file artifact being more than 1 MB. Even if it becomes 100 MB, we can handle that.

lambdageek committed:
better verbose output from ObjectFileScraper

lambdageek committed:
turn off whole program optimizations for data-descriptor.obj

On windows /GL creates object files that cdac-build-tool cannot read

It's ok to do this because we don't ship data-descriptor.obj as part
of the product - it's only used to generate the cDAC descriptor

lambdageek committed:
C++-ify and add real Thread offsets

lambdageek committed:
no C99 designated initializers in C++ until C++20

lambdageek committed:
build data descriptor after core runtime

lambdageek committed:
fix gcc build

lambdageek committed:
fix typo

lambdageek committed:
simplify ObjectFileScraper

just read the whole file into memory

lambdageek committed:
more dependencies

lambdageek committed:
try fix source build

lambdageek committed:
remove fixme; document

lambdageek committed:
Merge commit 'origin/main^' into cdac-contract-tool

lambdageek committed:
invoke 'dotnet cmake-build-tool.dll' instead of 'dotnet run --project'

lambdageek committed:
clean up macro boilerplate

lambdageek committed:
platform flags

lambdageek committed:
turn off verbose output

lambdageek commented:
@AaronRobinsonMSFT @elinor-fung cdac-build-tool is ready for a review

lambdageek committed:
can't use constexpr function in coreclr

because debugreturn.h defines a `return` macro that expands to
something that is not c++11 constexpr

lambdageek commented:
/azp run help

lambdageek commented:
/azp run runtime-community

elinor-fung reviewed (commented):
Were there steps/tools you found helpful for investigating issues during development? I see the tool has a verbose option that I assume should aid investigation.

There's a lot of moving parts here and my concern is the 'I specified something wrong, now what?' scenario. If there are specific things you found useful, I think having some sort of comments or doc around investigating issues would be nice.

For example, I have a hard time reading through and tracking the macros in `data-descriptor.cpp`. Could preprocessor output - either having it to a file by default or documenting how to get it - be useful for diagnosing problems?

lambdageek committed:
Code review fixups

Co-authored-by: Elinor Fung <elfung@microsoft.com>

lambdageek committed:
Rename "aux data" to "pointer data"

lambdageek committed:
rename "data-descriptor" to "datadescriptor"

lambdageek committed:
simplify linking

lambdageek committed:
cdac-build-tool don't build dotnet tool; turn on analyzers

lambdageek committed:
update README

lambdageek committed:
rationalize naming; update docs; add some inline comments

lambdageek commented on src/coreclr/tools/cdac-build-tool/data-descriptor.md:
```diff
@@ -0,0 +1,177 @@
+
+The blob begins with a directory that gives the relative offsets of the `Types`, `FieldPool`,
+`GlobalLiteralValues`, `GlobalPointerValues` and `Names` fields of the blob.  The number of elements of each of the arrays is
+next. This is followed by the sizes of the spec structs.
```
@elinor-fung  I added some comments in the "production" `datadescriptor.cpp` file, but in retrospect I think you were asking for something else here and I'm not 100% sure I get what you're after.  Do you just want me to move some of the text in this markdown doc next into the code block?  And I guess you're also asking for the definitions of some of the auxiliary structs like `FieldSpec` and `GlobalLiteralSpec` etc?

AaronRobinsonMSFT commented on src/coreclr/debug/runtimeinfo/contractpointerdata.cpp:
```diff
@@ -0,0 +1,22 @@
+extern "C"
+{
+  
+extern const uintptr_t contractDescriptorPointerData[];
```
Is this declaration needed?

lambdageek replied:
interestingly, yes.  without it clang doesn't emit the constant into the object file at all

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ComposeCommand.cs:
```diff
@@ -0,0 +1,92 @@
+        Add(contractFile);
+        SetAction(Run);
+    }
+
```
```suggestion
```

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ContractReader.cs:
```diff
@@ -0,0 +1,64 @@
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+
```
```suggestion
```

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ContractReader.cs:
```diff
@@ -0,0 +1,64 @@
+
+public partial class ContractReader
+{
+
```
```suggestion
```

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ContractReader.cs:
```diff
@@ -0,0 +1,64 @@
+        _builder = builder;
+    }
+
+    public async Task<bool> ParseContracts(string contractFilePath, CancellationToken token = default)
```
Do we need the `bool`? It seems like we are relying on exceptions for the failure case.

lambdageek replied:
removed the exceptions.  added line numbers to error messages

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ObjectFileScraper.cs:
```diff
@@ -0,0 +1,497 @@
+        public bool LittleEndian { get; }
+        private long _position;
+
+        public long MagicStart => HeaderStart - MagicLE.Length;
```
I assume that both LE and BE are the same length. I would add a comment this is intentional.

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ObjectFileScraper.cs:
```diff
@@ -0,0 +1,497 @@
+            _position = headerStart;
+        }
+
+        public static bool FindMagic(ReadOnlyMemory<byte> bytes, [NotNullWhen(true)] out ScraperState? scraperState)
```
```suggestion
        public static bool CreateScaperState(ReadOnlyMemory<byte> bytes, [NotNullWhen(true)] out ScraperState? scraperState)
```

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ObjectFileScraper.cs:
```diff
@@ -0,0 +1,497 @@
+        public uint GetUInt32(long offset) => LittleEndian ? BinaryPrimitives.ReadUInt32LittleEndian(Data.Span.Slice((int)offset)) : BinaryPrimitives.ReadUInt32BigEndian(Data.Span.Slice((int)offset));
+        public ushort GetUInt16(long offset) => LittleEndian ? BinaryPrimitives.ReadUInt16LittleEndian(Data.Span.Slice((int)offset)) : BinaryPrimitives.ReadUInt16BigEndian(Data.Span.Slice((int)offset));
+        public byte GetByte(long offset) => Data.Span[(int)offset];
+
```
```suggestion
```

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ObjectFileScraper.cs:
```diff
@@ -0,0 +1,497 @@
+        public byte GetByte(long offset) => Data.Span[(int)offset];
+
+
+        public void GetBytes(long offset, Span<byte> buffer) => Data.Span.Slice((int)offset, buffer.Length).CopyTo(buffer);
```
```suggestion
        public void GetBytes(long offset, ref Span<byte> buffer) => Data.Span.Slice((int)offset, buffer.Length).CopyTo(buffer);
```

Is this correct without the `ref`?

lambdageek replied:
> Is this correct without the `ref`?

Yes. `buffer` is a pair of a pointer and a length. `Span<T>.CopyTo()` doesn't modify where `buffer` points or how big it is.

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ObjectFileScraper.cs:
```diff
@@ -0,0 +1,497 @@
+            _position += count;
+        }
+    }
+
```
```suggestion
```

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ObjectFileScraper.cs:
```diff
@@ -0,0 +1,497 @@
+    }
+
+
+    private struct HeaderDirectory
```
Link to the specification for this data structure?

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/ObjectFileScraper.cs:
```diff
@@ -0,0 +1,497 @@
+
+    private static void DumpHeaderDirectory(HeaderDirectory headerDirectory)
+    {
+        Console.WriteLine($"Baseline Start = {headerDirectory.FlagsAndBaselineStart}");
+        Console.WriteLine($"Types Start = {headerDirectory.TypesStart}");
+        Console.WriteLine($"Field Pool  Start = {headerDirectory.FieldPoolStart}");
+        Console.WriteLine($"Global Literals Start = {headerDirectory.GlobalLiteralValuesStart}");
+        Console.WriteLine($"Global Pointers Start = {headerDirectory.GlobalPointersStart}");
+        Console.WriteLine($"Names Start = {headerDirectory.NamesStart}");
+        Console.WriteLine($"TypeCount = {headerDirectory.TypeCount}");
+        Console.WriteLine($"FieldPoolCount = {headerDirectory.FieldPoolCount}");
+        Console.WriteLine($"GlobalLiteralValuesCount = {headerDirectory.GlobalLiteralValuesCount}");
+        Console.WriteLine($"GlobalPointerValuesCount = {headerDirectory.GlobalPointerValuesCount}");
```
Nit: It looks like some of these are offsets. Offset in decimal form are very hard to reason about. I think it might be easier to read as hexadecimal if they are nicely aligned it creates easily readable output.

AaronRobinsonMSFT commented on src/coreclr/vm/cdac.h:
```diff
@@ -0,0 +1,22 @@
+
```
Header.

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/README.md:
````diff
@@ -0,0 +1,133 @@
+Running the `cdac-build-tool` with the `-v` verbose option will show progress
+
+```console
+$ ../../../../dotnet.sh run --project cdac-build-tool.csproj -- compose -v -o /tmp/contract.c /tmp/sample.o
````
I would go in the other direction. Assume this is at the room and point to the `csproj`. It is must easier to look for specific paths to update as opposed to arbitrary parent directory chains.

AaronRobinsonMSFT commented on src/coreclr/tools/cdac-build-tool/DataDescriptorModel.cs:
```diff
@@ -0,0 +1,445 @@
+            {
+                Console.WriteLine($"  Field: {fieldName}");
+                Console.WriteLine($"    Type: {field.Type}");
+                Console.WriteLine($"    Offset: {field.Offset}");
```
I think offsets are clearer as hexadecimal.

lambdageek commented on src/coreclr/vm/cdac.h:
```diff
@@ -0,0 +1,22 @@
+
+#ifndef CDAC_H__
```
I'm going to rename this file.  We already have another `cdac.h` in a different directory in #100946 - too confusing.

lambdageek committed:
rm whitespace

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

lambdageek committed:
renamce cdac.h to cdacoffsets.h

lambdageek committed:
comments

lambdageek committed:
fixup README

lambdageek committed:
hex offsets; improved formatting

lambdageek committed:
don't throw in ParseContracts; add line numbers to errors

lambdageek committed:
spelling

lambdageek committed:
Update README.md

lambdageek commented on src/coreclr/tools/cdac-build-tool/DataDescriptorModel.cs:
```diff
@@ -0,0 +1,446 @@
+        }
+    }
+
+    public string ToJson()
```
Because we're doing things in a slightly weird way for the "compact" format, we will need custom serializers. At that point it's just more boilerplate to get to this same place, imo.

lambdageek commented on src/coreclr/debug/runtimeinfo/datadescriptor.h:
```diff
@@ -0,0 +1,139 @@
+#define CDAC_GLOBALS_END()
+#endif
+
+CDAC_BASELINE("empty")
```
It's hard to have completely optional data with the preprocessor since everything here expands to a single global initialization expression. I would need some keyword for "no baseline".

lambdageek replied:
Follow-up work: https://github.com/dotnet/runtime/issues/101257

lambdageek commented on src/coreclr/debug/runtimeinfo/datadescriptor.cpp:
```diff
@@ -0,0 +1,301 @@
+            /* .GlobalLiteralSpecSize = */ sizeof(struct GlobalLiteralSpec),
+            /* .GlobalPointerSpecSize = */ sizeof(struct GlobalPointerSpec),
+        },
+        /* .PlatformFlags = */ (sizeof(void*) == 4 ? 0x02 : 0) | 0x01,
```
I can get rid of it.  I didn't want a 0 byte in the contract descriptor to be a valid flags value... I'm not sure there was more of a reason than that.

lambdageek replied:
Follow-up work: https://github.com/dotnet/runtime/issues/101257

lambdageek committed:
change input format for contracts to jsonc

lambdageek committed:
add custom JsonConverter instances for the compact json representation

lambdageek committed:
code review changes

lambdageek committed:
inline comments into the data-descriptor-blob C code

lambdageek committed:
cleanup

lambdageek committed:
comments and markdown

lambdageek commented on src/coreclr/tools/cdac-build-tool/ContractDescriptorSourceFileEmitter.cs:
```diff
@@ -0,0 +1,84 @@
+
+    public void SetPointerDataCount(int count)
+    {
+        Elements[PointerDataCount] = count.ToString();
```
yes.  I'll add it in `DataDescriptorModel.PointerDataCount`

lambdageek commented on src/coreclr/debug/runtimeinfo/datadescriptor.h:
```diff
@@ -0,0 +1,139 @@
+
+CDAC_TYPES_END()
+
+CDAC_GLOBALS_BEGIN()
```
I think their main use right now is just to group all the types and all the globals into separate collections.  I'm not sure if anything would go wrong if things were arranged differently, but since I didn't explicitly set things up to allow interleaving types and globals, I'm not sure what will happen. (the likeliest suspect is the names pool)

lambdageek committed:
Apply suggestions from code review

Co-authored-by: Elinor Fung <elfung@microsoft.com>

lambdageek committed:
simplify; bug fix - PointerDataCount include placeholder

lambdageek committed:
fixup docs

lambdageek committed:
Merge remote-tracking branch 'origin/main' into cdac-contract-tool

lambdageek committed:
Update src/coreclr/tools/cdac-build-tool/JsonConverter/GlobalModelJsonConverter.cs

Co-authored-by: Elinor Fung <elfung@microsoft.com>

lambdageek committed:
one more set of feedback changes

lambdageek commented:
All the windows build lanes are failing with something like:

```
.dotnet\sdk\9.0.100-preview.3.24204.13\Microsoft.Common.CurrentVersion.targets(4806,5): error MSB3026: (NETCORE_ENGINEERING_TELEMETRY=Build) Could not copy "D:\a\_work\1\s\artifacts\obj\ILLink.Tasks\Debug\net9.0\ILLink.Tasks.dll" to "D:\a\_work\1\s\artifacts\bin\ILLink.Tasks\Debug\net9.0\ILLink.Tasks.dll". Beginning retry 1 in 1000ms. The process cannot access the file 'D:\a\_work\1\s\artifacts\bin\ILLink.Tasks\Debug\net9.0\ILLink.Tasks.dll' because it is being used by another process. The file is locked by: ".NET Host (5476)"
```

which looks like we're running the in-tree `artifacts/bin/.../ILLink.Tasks.dll` at the same time as we're building it (a second time?)  I guess the obvious suspect is the compilation of `cdac-build-tool`, although i'm unclear why that would use the ILLink tasks from the in-tree build

lambdageek committed:
maybe no live ILLink ?

shushanhf commented:
After this PR, there is an error during coreclr camke-config step by command
```shell
####linux-debian
./build-runtime.sh -debug -loongarch64 -nopgooptimize -skipmanaged

-- Looking for process_vm_readv
-- Looking for process_vm_readv - found
CMake Error at debug/runtimeinfo/CMakeLists.txt:57 (message):
  No cdac-build-tool set or does not exist


-- Configuring incomplete, errors occurred!
See also "/home/qiao/work_qiao/runtime/artifacts/obj/coreclr/linux.loongarch64.Debug/CMakeFiles/CMakeOutput.log".
See also "/home/qiao/work_qiao/runtime/artifacts/obj/coreclr/linux.loongarch64.Debug/CMakeFiles/CMakeError.log".
```

am11 commented on src/coreclr/debug/runtimeinfo/CMakeLists.txt:
```diff
@@ -37,3 +37,46 @@ endif()
+
+set(GENERATED_CDAC_DESCRIPTOR_DIR "${CMAKE_CURRENT_BINARY_DIR}/cdac")
+set(CONTRACT_DESCRIPTOR_OUTPUT "${GENERATED_CDAC_DESCRIPTOR_DIR}/contract-descriptor.c")
+if("${CDAC_BUILD_TOOL_BINARY_PATH}" STREQUAL "" OR NOT EXISTS "${CDAC_BUILD_TOOL_BINARY_PATH}")
```
nit:

```diff
- "${CDAC_BUILD_TOOL_BINARY_PATH}" STREQUAL ""
+ NOT CDAC_BUILD_TOOL_BINARY_PATH
```
it checks both "not defined" and "evaluates to empty string".

am11 commented on src/coreclr/debug/runtimeinfo/CMakeLists.txt:
```diff
@@ -37,3 +37,46 @@ endif()
+add_custom_command(
+    OUTPUT "${CONTRACT_DESCRIPTOR_OUTPUT}"
+    VERBATIM
+    COMMAND dotnet ${CDAC_BUILD_TOOL_BINARY_PATH} compose -o "${CONTRACT_DESCRIPTOR_OUTPUT}" -c "${CONTRACT_FILE}" $<TARGET_OBJECTS:cdac_data_descriptor>
```
1. `src/coreclr/build-runtime.sh` direct invocation skips requiring dotnet/msbuild, which is necessary for new platform port work. This PR assumes dotnet/msbuild are always available. We should handle that case.

2. Secondly, can use published cdac-build-tool instead of assembly so instead of requiring `dotnet` in PATH, it uses executable (AOT'd, singilefilehost, apphost -based doesn't matter)? If that's not possible, then please pass `DOTNET_HOST_PATH` (set by SDK to the executing msbuild) via cmakeargs as well.

lambdageek replied:
Can we just make embedding the contract descriptor optional if someone is doing a build without msbuild?

lambdageek replied:
> If making it optional doesn't affect debugging, that's good

in the future, it will make using something like dotnet-sos impossible

I don't understand what scenarios don't have a .NET sdk available.  any new platform bringup will start with cross compiling from an existing platform with a .NET sdk and a cross compiler toolchain, right?  that case should be handled by the cmake+msbuild infrastructure


I added https://github.com/dotnet/runtime/pull/101297

---

> ```diff
> -    COMMAND dotnet ${CDAC_BUILD_TOOL_BINARY_PATH} compose -o "${CONTRACT_DESCRIPTOR_OUTPUT}" -c "${CONTRACT_FILE}" $<TARGET_OBJECTS:cdac_data_descriptor>
> +    COMMAND "${CLR_DOTNET_HOST_PATH}" "${CDAC_BUILD_TOOL_BINARY_PATH}" compose -o "${CONTRACT_DESCRIPTOR_OUTPUT}" -c "${CONTRACT_FILE}" $<TARGET_OBJECTS:cdac_data_descriptor>
> ```

I had something like this before and it broke source build because it was pointing to the repo root `.dotnet/dotnet`, not to `artifacts/sb/.dotnet/dotnet`.  AFAICT it's fine to just call "dotnet" from the path.  eng/commons/build.sh alaways adds the right SDK dir to the path - it's done by InitializeDotNetCli, and it is source-build aware

lambdageek replied:
*Update* I'm educating myself over in https://github.com/dotnet/installer/pull/19534#issuecomment-2066812161

I think previously I used something other than `DOTNET_HOST_PATH`

lambdageek replied:
Updated https://github.com/dotnet/runtime/pull/101297 to use DOTNET_HOST_PATH if it's set
